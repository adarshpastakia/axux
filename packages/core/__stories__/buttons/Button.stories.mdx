import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title
} from "@storybook/addon-docs";
import { AxButton, COLOR } from "../../src";
import {
  ButtonStory,
  OnlyIconStory,
  WithBadgeStory,
  WithIconStory,
  WithTooltipStory
} from "./ButtonStory";
import { mdiBell } from "@mdi/js";

<Meta
  title="@core/Buttons/Button"
  component={AxButton}
  parameters={{ jest: ["Button.test.tsx"] }}
/>

<Anchor storyId="core-buttons-button--button">
  <Title>Button</Title>
</Anchor>

<Description of={AxButton} />

<Canvas>
  <Story name="Button">{ButtonStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--with-icon">
  <Heading>With Icon</Heading>
</Anchor>

<Canvas>
  <Story name="With Icon">{WithIconStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--only-icon">
  <Heading>Icon Only</Heading>
</Anchor>

<Canvas>
  <Story name="Only Icon">{OnlyIconStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--with-tooltip">
  <Heading>With Tooltip</Heading>
</Anchor>

<Canvas>
  <Story name="With Tooltip">{WithTooltipStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--with-badge">
  <Heading>With Badge</Heading>
</Anchor>

<Canvas>
  <Story name="With Badge">{WithBadgeStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--styles">
  <Heading>Styles</Heading>
</Anchor>

<Canvas>
  <Story name="Styles">
    <div>
      <AxButton type="default">Default</AxButton>
      <AxButton type="outline">Outline</AxButton>
      <AxButton type="solid">Solid</AxButton>
      <AxButton type="link">Link</AxButton>
    </div>
    <div>
      <AxButton data-popover="true" type="default">
        Default
      </AxButton>
      <AxButton data-popover="true" type="outline">
        Outline
      </AxButton>
      <AxButton data-popover="true" type="solid">
        Solid
      </AxButton>
      <AxButton data-popover="true" type="link">
        Link
      </AxButton>
    </div>
    <div>
      <AxButton split data-popover="true" type="default">
        Default
      </AxButton>
      <AxButton split data-popover="true" type="outline">
        Outline
      </AxButton>
      <AxButton split data-popover="true" type="solid">
        Solid
      </AxButton>
      <AxButton split data-popover="true" type="link">
        Link
      </AxButton>
    </div>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--icon-styles">
  <Heading>Icon Styles</Heading>
</Anchor>

<Canvas>
  <Story name="Icon Styles">
    <AxButton icon={mdiBell}>Button</AxButton>
    <AxButton icon={mdiBell} iconAlign="end">
      Button
    </AxButton>
    <AxButton icon={mdiBell} iconHilight>
      Button
    </AxButton>
    <AxButton icon={mdiBell} iconHilight iconAlign="end">
      Button
    </AxButton>
    <AxButton type="solid" icon={mdiBell} iconHilight>
      Button
    </AxButton>
    <AxButton type="solid" icon={mdiBell} iconHilight iconAlign="end">
      Button
    </AxButton>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--sizes">
  <Heading>Sizes</Heading>
</Anchor>

<Canvas>
  <Story name="Sizes">
    <AxButton size="sm">Small</AxButton>
    <AxButton size="default">Default</AxButton>
    <AxButton size="md">Medium</AxButton>
    <AxButton size="lg">Large</AxButton>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--themes">
  <Heading>Themes</Heading>
</Anchor>

<Canvas>
  <Story name="Themes">
    <div>
      {Object.values(COLOR).map((color) => (
        <AxButton key={color} color={color}>
          {color}
        </AxButton>
      ))}
    </div>
    <div>
      {Object.values(COLOR).map((color) => (
        <AxButton key={color} type="solid" color={color}>
          {color}
        </AxButton>
      ))}
    </div>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--variants">
  <Heading>Variants</Heading>
</Anchor>

<Canvas>
  <Story name="Variants">
    <AxButton.Positive>Positive Button</AxButton.Positive>
    <AxButton.Negative>Negative Button</AxButton.Negative>
    <AxButton.Neutral>Neutral Button</AxButton.Neutral>
    <AxButton.Action message="Done!">Action Button</AxButton.Action>
    <AxButton.Confirm message="Are you sure?">Confirm Button</AxButton.Confirm>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--states">
  <Heading>States</Heading>
</Anchor>

<Canvas>
  <Story name="States">
    <AxButton>Normal</AxButton>
    <AxButton isLoading>Loading</AxButton>
    <AxButton isDisabled>Disabled</AxButton>
  </Story>
</Canvas>

<Heading>Props</Heading>

<ArgsTable of={AxButton} sort="requiredFirst" />
