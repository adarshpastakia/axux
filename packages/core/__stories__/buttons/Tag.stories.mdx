import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title
} from "@storybook/addon-docs";
import { AxFlexBox, AxHotKeyLabel, AxTag, COLOR, PALETTE } from "../../src";
import {
  Clickable,
  Hotkeys, Removable,
  TagSizes,
  TagStory,
  TagThemes,
  WithBadgeStory,
  WithIconStory,
  WithTooltipStory
} from "./TagStory";
import { mdiBell } from "@mdi/js";

<Meta title="@core/Buttons/Tag" component={AxTag} parameters={{ jest: ["Tag.test.tsx"] }} />

<Anchor storyId="core-buttons-tag--tag">
  <Title>Tag</Title>
</Anchor>

<Description of={AxTag} />

<Canvas>
  <Story name="Tag">{TagStory}</Story>
  <Story name="With Icon">{WithIconStory}</Story>
  <Story name="With Tooltip">{WithTooltipStory}</Story>
  <Story name="With Badge">{WithBadgeStory}</Story>
  <Story name="Clickable">{Clickable}</Story>
  <Story name="Removable">{Removable}</Story>
</Canvas>

<Anchor storyId="core-buttons-tag--hotkey">
  <Heading>Hotkey Label</Heading>
</Anchor>

<Canvas>
  <Story name="Hotkey">{Hotkeys.bind({})}</Story>
</Canvas>

<Anchor storyId="core-buttons-tag--sizes">
  <Heading>Sizes</Heading>
</Anchor>

<Canvas>
  <Story name="Sizes">{TagSizes.bind({})}</Story>
</Canvas>

<Anchor storyId="core-buttons-tag--themes">
  <Heading>Themes</Heading>
</Anchor>

<Canvas>
  <Story name="Themes">{TagThemes.bind({})}</Story>
</Canvas>

<Heading>Props</Heading>

<ArgsTable components={{ AxTag, AxHotKeyLabel }} sort="requiredFirst" />
