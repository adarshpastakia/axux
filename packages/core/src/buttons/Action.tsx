/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import { isString } from "@axux/utilities/src";
import { useCallback, useEffect, useMemo, useState, type FC } from "react";
import { AxAnimation } from "../animations";
import { AxTooltip } from "../overlays/Tooltip";
import { AxButton, type ButtonProps } from "./Button";

export interface ActionProps extends ButtonProps {
  /**
   * action message
   */
  message: string;
  /**
   * action type
   */
  actionType?: "danger" | "success";
}

/**
 * Action button to show tooltip message on success
 *
 * @requires This button expects a `onClick` return value of boolean type,
 * if true will display given message and display animated check/cross
 *
 * @prop message
 * @prop actionType - danger | success
 * @prop see Button
 */
export const ActionButton: FC<ActionProps> = ({
  onClick,
  message,
  color,
  variant: style,
  tooltip,
  actionType = "success",
  children,
  ...props
}) => {
  const [actionDone, setActionDone] = useState(false);

  /** ***************** handle click *******************/
  const doClick = useCallback(() => {
    const ret = onClick?.();
    void Promise.resolve(ret).then((b) => {
      if (b !== false) {
        setActionDone(true);
      }
    });
  }, [onClick]);

  /** ***************** reset done state on timeout *******************/
  useEffect(() => {
    if (actionDone) {
      const timer = setTimeout(() => setActionDone(false), 2000);
      return () => clearTimeout(timer);
    }
  }, [actionDone]);

  /** ***************** normalize tooltip content *******************/
  const tooltipContent = useMemo(() => {
    if (actionDone || !tooltip) return { content: message };
    return isString(tooltip) ? { content: tooltip } : tooltip;
  }, [message, tooltip, actionDone]);

  /** ***************** component *******************/
  return (
    <AxTooltip
      {...tooltipContent}
      isDisabled={!tooltip}
      isOpen={actionDone || undefined}
      placement="top"
    >
      <AxButton
        {...props}
        onClick={doClick}
        data-active={actionDone}
        data-action-done={actionDone}
        color={actionDone ? actionType : color}
        variant={actionDone ? "solid" : style}
        data-extra={
          actionDone && (
            <div className="animated-svg">
              {actionType === "success" && (
                <AxAnimation.Check className="text-success-base" />
              )}
              {actionType === "danger" && (
                <AxAnimation.Cross className="text-danger-base" />
              )}
            </div>
          )
        }
      >
        {children}
      </AxButton>
    </AxTooltip>
  );
};
