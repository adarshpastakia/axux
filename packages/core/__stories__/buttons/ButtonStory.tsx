// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { action } from "@storybook/addon-actions";
import { Story } from "@storybook/react";
import { PropsWithChildren } from "react";
import { StateLabel } from "../../../../storybook/components";
import { AxButton, AxFlexBox, AxMenu, COLOR } from "../../src";
import { ButtonProps } from "../../src/buttons/Button";

export const ButtonTemplate: Story<PropsWithChildren<ButtonProps>> = (props) => (
  <AxButton {...props} />
);

export const ButtonStyles: Story<Omit<ButtonProps, "type">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      {["default", "outline", "solid", "link"].map((type: AnyObject) => (
        <AxFlexBox.Col flex="auto" key={type}>
          <StateLabel>{type}</StateLabel>
          <AxButton {...props} type={type}>
            Click Me!
          </AxButton>
        </AxFlexBox.Col>
      ))}
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonSizes: Story<Omit<ButtonProps, "size">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      {[
        ["sm", "Small"],
        ["default", "Default"],
        ["md", "Medium"],
        ["lg", "Large"]
      ].map(([size, label]: AnyObject) => (
        <AxFlexBox.Col flex="auto" key={size}>
          <StateLabel>{label}</StateLabel>
          <AxButton {...props} size={size}>
            Click Me!
          </AxButton>
        </AxFlexBox.Col>
      ))}
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonThemes: Story<Omit<ButtonProps, "color">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      {Object.values(COLOR).map((color) => (
        <AxFlexBox.Col flex="auto" key={color}>
          <StateLabel>{color}</StateLabel>
          <AxButton {...props} color={color}>
            {color}
          </AxButton>
        </AxFlexBox.Col>
      ))}
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const DropdownStory: Story<Omit<ButtonProps, "split" | "children">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      <AxFlexBox.Col flex="auto">
        <AxButton.Dropdown label="Click Me!" {...props} split={false}>
          <AxMenu>
            <AxMenu.Item label="Dropdown menu" />
          </AxMenu>
        </AxButton.Dropdown>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <AxButton.Dropdown label="Click Me!" {...props} split="Button dropdown">
          <AxMenu>
            <AxMenu.Item label="Dropdown menu" />
          </AxMenu>
        </AxButton.Dropdown>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const IconStyles: Story<Omit<ButtonProps, "icon" | "iconAlign" | "iconHilight">> = (
  props
) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Align start</StateLabel>
        <AxButton {...props} icon="mdi mdi-bell">
          Button
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Align end</StateLabel>
        <AxButton {...props} icon="mdi mdi-bell" iconAlign="end">
          Button
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Hilight end</StateLabel>
        <AxButton {...props} icon="mdi mdi-bell" iconHilight>
          Button
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Hilight start</StateLabel>
        <AxButton {...props} icon="mdi mdi-bell" iconHilight iconAlign="end">
          Button
        </AxButton>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonVariants: Story<ButtonProps> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Positive</StateLabel>
        <AxButton.Positive {...props}>Click Me!</AxButton.Positive>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Negative</StateLabel>
        <AxButton.Negative {...props}>Click Me!</AxButton.Negative>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Neutral</StateLabel>
        <AxButton.Neutral {...props}>Click Me!</AxButton.Neutral>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Action</StateLabel>
        <AxButton.Action {...props} message="Action completed!">
          Click Me!
        </AxButton.Action>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Confirm</StateLabel>
        <AxButton.Confirm {...props} message="Confirm action?" onClick={action("Confirm:onClick")}>
          Click Me!
        </AxButton.Confirm>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonStates: Story<Omit<ButtonProps, "isLoading" | "isDisabled">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Normal</StateLabel>
        <AxButton {...props}>Click Me!</AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Loading</StateLabel>
        <AxButton {...props} isLoading>
          Click Me!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Disabled</StateLabel>
        <AxButton {...props} isDisabled>
          Click Me!
        </AxButton>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonStory = ButtonTemplate.bind({});
ButtonStory.args = {
  children: "Click Me!"
};

export const WithIconStory = ButtonTemplate.bind({});
WithIconStory.args = {
  icon: "mdi mdi-bell",
  children: "Click Me!"
};

export const OnlyIconStory = ButtonTemplate.bind({});
OnlyIconStory.args = {
  "aria-label": "Icon button",
  icon: "mdi mdi-bell"
};

export const WithTooltipStory = ButtonTemplate.bind({});
WithTooltipStory.args = {
  children: "Tooltip!",
  tooltip: {
    placement: "bottom",
    content: "This is a clickable button"
  }
};

export const WithBadgeStory = ButtonTemplate.bind({});
WithBadgeStory.args = {
  children: "Click Me!",
  badge: {
    color: "danger",
    value: 9
  }
};

export default { title: "Example/Button", component: AxButton };
