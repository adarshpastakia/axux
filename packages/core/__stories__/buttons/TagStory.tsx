// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { Story } from "@storybook/react";
import { PropsWithChildren } from "react";
import { AxTag, TagProps } from "../../src/buttons/Tag";

const Template: Story<PropsWithChildren<TagProps>> = (props) => <AxTag {...props} />;

export const TagStory = Template.bind({});
TagStory.args = {
  children: "Tag"
};

export const WithIconStory = Template.bind({});
WithIconStory.args = {
  icon: "mdi mdi-bell",
  children: "Tag"
};

export const WithTooltipStory = Template.bind({});
WithTooltipStory.args = {
  children: "Tag",
  tooltip: {
    placement: "bottom",
    content: "Test tooltip"
  }
};

export const WithBadgeStory = Template.bind({});
WithBadgeStory.args = {
  children: "Tag",
  badge: 9
};

export default { title: "Example/Tag", component: AxTag };
