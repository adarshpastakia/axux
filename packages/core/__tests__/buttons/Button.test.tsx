// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { composeStories } from "@storybook/testing-react";
import { act, fireEvent, render, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import * as stories from "../../__stories__/buttons/ButtonStory";

const { ButtonStory, WithIconStory, OnlyIconStory, WithTooltipStory } = composeStories(stories);

describe("Button", () => {
  it("renders basic button", (done) => {
    const fragment = render(<ButtonStory />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector(".ax-button__icon")).toBeNull();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders with icon", (done) => {
    const fragment = render(<WithIconStory />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector(".ax-button__icon")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders only icon", (done) => {
    const fragment = render(<OnlyIconStory />);
    expect(fragment.container).not.toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector(".ax-button__icon")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders color", (done) => {
    const fragment = render(<ButtonStory color="secondary" />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector("[data-theme='secondary']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    done();
  });

  it("renders style", (done) => {
    const fragment = render(<ButtonStory type="outline" />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector("[data-style='outline']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders size", (done) => {
    const fragment = render(<ButtonStory size="md" />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector("[data-size='md']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders block", (done) => {
    const fragment = render(<ButtonStory block />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector("[data-block='true']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders icon align", (done) => {
    const fragment = render(<WithIconStory iconAlign="end" />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector("[data-icon-align='end']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders icon hilight", (done) => {
    const fragment = render(<WithIconStory iconHilight />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector("[data-icon-hilight='true']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders link", (done) => {
    const fragment = render(<WithIconStory href="/test" />, {
      wrapper: BrowserRouter
    });
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector("[href='/test']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders new tab link", (done) => {
    const fragment = render(<WithIconStory href="/test" target="_blank" />, {
      wrapper: BrowserRouter
    });
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector("[target='_blank']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders with spinner", (done) => {
    const fragment = render(<WithIconStory isLoading />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector(".ax-button__spinner")).toBeInTheDocument();
    expect(fragment.container.querySelector("[data-busy='true']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    done();
  });

  it("renders disabled", (done) => {
    const fragment = render(<WithIconStory isDisabled />);
    expect(fragment.container).toHaveTextContent("Click Me!");
    expect(fragment.container.querySelector("[data-disabled='true']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders with tooltip", async () => {
    const fragment = render(<WithTooltipStory />);
    expect(fragment.container).toHaveTextContent("Tooltip!");

    const button = fragment.container.querySelector(".ax-button__inner") as HTMLElement;
    act(() => {
      fireEvent.mouseOver(button);
    });
    await waitFor(() => fragment.container.querySelector(".ax-tooltip"));
    expect(fragment.container).toHaveTextContent("This is a clickable button");
    expect(fragment.container).toMatchSnapshot();
    act(() => {
      fireEvent.mouseOut(button);
    });
    expect(fragment.container).not.toHaveTextContent("This is a clickable button");
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
  });

  it("should trigger click event", (done) => {
    const handler = jest.fn();
    const fragment = render(<ButtonStory onClick={handler}>Click</ButtonStory>);
    fireEvent.click(fragment.getByText("Click"));
    expect(handler).toBeCalled();
    fragment.unmount();
    done();
  });
});
