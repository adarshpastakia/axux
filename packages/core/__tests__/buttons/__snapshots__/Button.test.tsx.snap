// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders basic button 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders block 1`] = `
<div>
  <div
    class="ax-button "
    data-block="true"
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders color 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="secondary"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders disabled 1`] = `
<div>
  <div
    class="ax-button "
    data-disabled="true"
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-icon ax-button__icon"
      >
        <i
          class="mdi mdi-bell"
        />
      </div>
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders icon align 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="end"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-icon ax-button__icon"
      >
        <i
          class="mdi mdi-bell"
        />
      </div>
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders icon hilight 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="true"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-icon ax-button__icon"
      >
        <i
          class="mdi mdi-bell"
        />
      </div>
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders link 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <a
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      href="/test"
    >
      <div
        class="ax-icon ax-button__icon"
      >
        <i
          class="mdi mdi-bell"
        />
      </div>
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </a>
  </div>
</div>
`;

exports[`Button renders new tab link 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="default"
    target="_blank"
  >
    <a
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      href="/test"
    >
      <div
        class="ax-icon ax-button__icon"
      >
        <i
          class="mdi mdi-bell"
        />
      </div>
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </a>
  </div>
</div>
`;

exports[`Button renders only icon 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="true"
      type="button"
    >
      <div
        class="ax-icon ax-button__icon"
      >
        <svg
          viewBox="0 0 24 24"
        >
          <path
            d="M21,19V20H3V19L5,17V11C5,7.9 7.03,5.17 10,4.29C10,4.19 10,4.1 10,4A2,2 0 0,1 12,2A2,2 0 0,1 14,4C14,4.1 14,4.19 14,4.29C16.97,5.17 19,7.9 19,11V17L21,19M14,21A2,2 0 0,1 12,23A2,2 0 0,1 10,21"
            fill="currentColor"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders size 1`] = `
<div>
  <div
    class="ax-button "
    data-size="md"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders style 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="outline"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders with icon 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-icon ax-button__icon"
      >
        <i
          class="mdi mdi-bell"
        />
      </div>
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders with spinner 1`] = `
<div>
  <div
    class="ax-button "
    data-busy="true"
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-icon ax-button__icon"
      >
        <i
          class="mdi mdi-bell"
        />
      </div>
      <div
        class="ax-icon ax-button__spinner ax-anim--spin"
      >
        <svg
          viewBox="0 0 24 24"
        >
          <path
            d="M24 12c-0.030-1.567-0.37-3.129-0.998-4.558-0.626-1.43-1.534-2.725-2.649-3.795s-2.436-1.918-3.867-2.476c-1.43-0.561-2.967-0.829-4.486-0.796-1.519 0.030-3.031 0.36-4.414 0.969-1.384 0.607-2.638 1.488-3.673 2.568s-1.855 2.36-2.395 3.745c-0.542 1.384-0.8 2.872-0.767 4.342 0.030 1.471 0.351 2.933 0.941 4.27 0.588 1.338 1.441 2.551 2.487 3.552s2.284 1.793 3.624 2.314c1.339 0.523 2.776 0.771 4.199 0.739 1.423-0.030 2.835-0.341 4.127-0.912 1.293-0.569 2.464-1.394 3.43-2.406s1.731-2.209 2.233-3.502c0.305-0.784 0.513-1.603 0.622-2.433 0.029 0.002 0.059 0.003 0.088 0.003 0.828 0 1.5-0.672 1.5-1.5 0-0.042-0.002-0.084-0.006-0.125h0.006zM21.617 15.983c-0.55 1.247-1.347 2.377-2.324 3.309s-2.133 1.668-3.381 2.151c-1.248 0.485-2.585 0.714-3.911 0.682-1.327-0.030-2.639-0.322-3.84-0.855-1.201-0.531-2.289-1.301-3.187-2.243s-1.606-2.057-2.070-3.259c-0.466-1.202-0.685-2.489-0.653-3.768 0.031-1.279 0.312-2.541 0.826-3.696 0.512-1.155 1.254-2.202 2.162-3.066s1.981-1.544 3.138-1.989c1.156-0.447 2.394-0.656 3.624-0.624 1.23 0.031 2.443 0.303 3.552 0.798 1.11 0.493 2.115 1.207 2.944 2.081s1.481 1.905 1.908 3.016c0.428 1.111 0.628 2.298 0.596 3.48h0.006c-0.003 0.041-0.006 0.083-0.006 0.125 0 0.774 0.586 1.41 1.338 1.491-0.146 0.816-0.387 1.613-0.721 2.367z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button renders with tooltip 1`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
  <div
    class="ax-popper ax-root ax-tooltip ax-bg--darkest ax-color--empty "
    data-show="true"
    data-show-arrow="true"
    style="position: fixed; left: 0px; top: 0px; min-width: 0;"
  >
    <div
      class="ax-popper--body"
    >
      <span>
        Test tooltip
      </span>
    </div>
    <div
      class="ax-popper--arrow"
      style="position: absolute;"
    />
  </div>
</div>
`;

exports[`Button renders with tooltip 2`] = `
<div>
  <div
    class="ax-button "
    data-size="default"
    data-style="default"
    data-theme="default"
  >
    <button
      class="ax-button__inner"
      data-icon-align="start"
      data-icon-hilight="false"
      data-no-label="false"
      type="button"
    >
      <div
        class="ax-button__label"
      >
        <span>
          Button
        </span>
      </div>
    </button>
  </div>
</div>
`;
